components/st-microelectronics/stm-api.stanza

```diff
@@ -208,54 +224,75 @@ public defn connect-named-power (stm:JITXObject, power-port-name:String, vdd-net
 
       property(vdd-pin-list[0].gnd-ref) = gnd-pin-list[0]
 
-
 public defn connect-power (stm:JITXObject) :
+  connect-power(stm, "power", VDD, VSS)
+
+public defn connect-battery (stm:JITXObject, bat-net:PwrPinNames, vdd-net:PwrPinNames) :
   inside pcb-module :
-    val vdd-pin-list = power-pins(stm, VDD)
-    val gnd-pin-list = gnd-pins(stm, VSS)
-    port power : power
-    for p in vdd-pin-list do : 
-      net VDD (power.vdd p)
-    for p in gnd-pin-list do :
-       net GND (p, power.gnd)
-
-    val v-range = recommended-voltage(property(vdd-pin-list[0].power-pin))
-    if contains?(v-range, 3.3):
-      property(vdd-pin-list[0].power-request) = [3.3, 0.1, 3.3 * 0.03]
-    else if contains?(v-range, 2.5):
-      property(vdd-pin-list[0].power-request) = [2.5, 0.1, 2.5 * 0.03]
-    else if contains?(v-range, 1.8):
-      property(vdd-pin-list[0].power-request) = [1.8, 0.1, 1.8 * 0.03]
-    else if contains?(v-range, 1.2):
-      property(vdd-pin-list[0].power-request) = [1.2, 0.1, 1.2 * 0.03]
-
-    property(vdd-pin-list[0].gnd-ref) = gnd-pin-list[0]
-
-    ; TODO Add battery option to API, currently assumes none.
-    if find-pins?(stm, "VBAT") :
-      net (stm.VBAT power.vdd)
+  ; TODO Add battery option to API, currently assumes none.
+    val b-net = get-pin-list(stm, to-string(bat-net))
+    if not empty?(b-net) :
+      val p-pin = get-pin-list(stm, to-string(vdd-net))
+      if length(p-pin) > 0 :
+        net (b-net p-pin[0])
 
+public defn connect-battery (stm:JITXObject, bat-net:PwrPinNames) :
+  connect-battery(stm, bat-net, VDD)
 
-public defn connect-reset (stm:JITXObject) :
+public defn connect-battery (stm:JITXObject) :
+  connect-battery(stm, VBAT, VDD)
+
+public defn connect-reset (stm:JITXObject reset-net:ResetPinNames vdd-net:PwrPinNames gnd-net:GndPinNames reset-pullup:Double reset-cap:Double) :
```

Comment:
I like the idea of extending this! But we could go further, enable reset-pullup and reset-cap to be JITXObjects, then users can specify things like low leakage caps or packages, or things we may not even think about today.

Reply:
@electricfungi So I think that's a good idea. I tried the following (but it did not like the fact that I was calling the JITXObject as a function):
```
public defn connect-reset (stm:JITXObject reset-net:ResetPinNames vdd-net:PwrPinNames gnd-net:GndPinNames reset-pullup:JITXObject|Double reset-cap:JITXObject|Double) :
  inside pcb-module :
    pin reset
    val reset-pin-list = get-pin-list(stm, reset-net)
    if not empty?(reset-pin-list) :
      val power-pins = power-pins-list(stm, VDD)
      for p in reset-pin-list do :
        net (p reset)
        if not empty?(power-pins) :
          match(reset-pullup) :
            (r: Double)     : res-strap(power-pins[0], p, r)
            (i: JITXObject) : inst rp : i(power-pins[0], p)
          match(reset-cap) :
            (c: Double)     : cap-strap(power-pins[0], p, c)
            (i: JITXObject) : inst cp : i(power-pins[0], p)
```
How should I fix this -- am I correct in the approach?

Reply:
I have tried the following but the Instantiable cannot take any arguments.
```
public defn connect-reset (stm:JITXObject reset-net:ResetPinNames vdd-net:PwrPinNames gnd-net:GndPinNames reset-pullup:Instantiable|JITXObject|Double reset-cap:Instantiable|JITXObject|Double) :
  inside pcb-module :
    pin reset
    val reset-pin-list = get-pin-list(stm, reset-net)
    if not empty?(reset-pin-list) :
      val power-pins = power-pins-list(stm, VDD)
      for p in reset-pin-list do :
        net (p reset)
        if not empty?(power-pins) :
          match(reset-pullup) :
            (r: Double)       : res-strap(power-pins[0], p, r)
            (i: Instantiable) : inst rp : i
            (j: JITXObject)   : false
          match(reset-cap) :
            (c: Double)     : cap-strap(power-pins[0], p, c)
            (i: Instantiable) : inst cp : i
            (j: JITXObject)   : false
```

Reply:
@electricfungi @CuppoJava Here is my evolution of the reset-net function. Note that it still does not cover that many cases.
```
public defn connect-reset (stm:JITXObject reset-net:ResetPinNames vdd-net:PwrPinNames gnd-net:GndPinNames reset-pullup:Instantiable|JITXObject|Double reset-cap:Instantiable|JITXObject|Double) :
  inside pcb-module :
    pin reset
    val reset-pin-list = get-pin-list(stm, reset-net)
    if not empty?(reset-pin-list) :
      val power-pins = power-pins-list(stm, VDD)
      for p in reset-pin-list do :
        net (p reset)
        if not empty?(power-pins) :
          match(reset-pullup) :
            (r: Double)       : res-strap(power-pins[0], p, r)
            (i: Instantiable) : let :
                                  inst rp : i
                                  net (rp.p[0] power-pins[0])
                                  net (rp.p[1] p)
            (j: JITXObject)   : false
          match(reset-cap) :
            (c: Double)     : cap-strap(power-pins[0], p, c)
            (i: Instantiable) : let :
                                  inst cp : i
                                  net (cp.p[0] power-pins[0])
                                  net (cp.p[1] p)
            (j: JITXObject)   : false
```